src/sacn/mem.c

```diff
@@ -660,6 +769,306 @@ void remove_receiver_from_list(SacnRecvThreadContext* recv_thread_context, SacnR
+{
+  SacnSourceNetint* netint = ALLOC_SACN_SOURCE_NETINT();
+
+  if (netint)
+  {
+    netint->id = *id;
+
+    if (etcpal_rbtree_insert(&source->netints, netint) != kEtcPalErrOk)
+    {
+      FREE_SACN_SOURCE_NETINT(netint);
+      netint = NULL;
+    }
+  }
+
+  return netint;
+}
+
+// Needs lock
+etcpal_error_t lookup_source_state(sacn_source_t source, uint16_t universe, SacnSource** source_state,
+                                   SacnSourceUniverse** universe_state)
+{
+  etcpal_error_t result = kEtcPalErrOk;
+
+  SacnSource* my_source_state = NULL;
+  SacnSourceUniverse* my_universe_state = NULL;
+
+  // Look up the source state.
+  my_source_state = etcpal_rbtree_find(&sources, &source);
+
+  if (!my_source_state)
+    result = kEtcPalErrNotFound;
+
+  // Look up the universe state.
+  if ((result == kEtcPalErrOk) && universe_state)
+  {
+    my_universe_state = etcpal_rbtree_find(&my_source_state->universes, &universe);
+
+    if (!my_universe_state)
+      result = kEtcPalErrNotFound;
+  }
+
+  if (result == kEtcPalErrOk)
+  {
+    if (source_state)
+      *source_state = my_source_state;
+    if (universe_state)
+      *universe_state = my_universe_state;
+  }
+
+  return result;
+}
+
+// Needs lock
+etcpal_error_t lookup_unicast_dest(sacn_source_t source, uint16_t universe, const EtcPalIpAddr* addr,
+                                   SacnUnicastDestination** unicast_dest)
+{
+  // Look up universe
+  SacnSourceUniverse* universe_state = NULL;
+  etcpal_error_t result = lookup_source_state(source, universe, NULL, &universe_state);
+
+  // Validate
+  if ((result == kEtcPalErrOk) && (!addr))
+    result = kEtcPalErrSys;
+
+  // Look up unicast destination
+  SacnUnicastDestination* found = NULL;
+  if (result == kEtcPalErrOk)
+  {
+    found = etcpal_rbtree_find(&universe_state->unicast_dests, addr);
+
+    if (!found)
+      result = kEtcPalErrNotFound;
+  }
+
+  // Pass back to application
+  if ((result == kEtcPalErrOk) && unicast_dest)
+    *unicast_dest = found;
+
+  return result;
+}
+
+// Needs lock
+EtcPalRbTree* get_sacn_sources()
+{
+  return &sources;
+}
+
+// Needs lock
+void remove_sacn_source_netint(SacnSource* source, SacnSourceNetint** netint)
+{
+  if (etcpal_rbtree_remove(&source->netints, *netint) == kEtcPalErrOk)
+  {
+    FREE_SACN_SOURCE_NETINT(*netint);
+    *netint = NULL;
+  }
+}
+
+// Needs lock
+void remove_sacn_unicast_dest(SacnSourceUniverse* universe, SacnUnicastDestination** dest, EtcPalRbIter* unicast_iter)
+{
+  SacnUnicastDestination* dest_to_remove = *dest;
+  *dest = etcpal_rbiter_next(unicast_iter);
+  etcpal_rbtree_remove(&universe->unicast_dests, dest_to_remove);
+  FREE_SACN_UNICAST_DESTINATION(dest_to_remove);
+}
+
+// Needs lock
+void remove_sacn_source_universe(SacnSource* source, SacnSourceUniverse** universe, EtcPalRbIter* universe_iter)
+{
+  SacnSourceUniverse* universe_to_remove = *universe;
+  *universe = etcpal_rbiter_next(universe_iter);
```

Comment:
This won't work. etcpal_rbtree_remove() invalidates all iterators.

Reply:
@nwagner I may have made a wrong assumption here. I will get back to you

Reply:
@nwagner Yes, just confirmed my original comment was correct via a test. In the open-source red-black tree implementation we use, iterators save a traversal path from the root of the tree to the current node they are on. Removing an element can cause the tree to be rebalanced, which invalidates that path.

Reply:
Ok...  Can you update the rbtree documentation?  That's a nasty gotcha.
Adding doesn't rebalance the tree, then?

Reply:
Adding can definitely rebalance the tree, yes. I would expect any operation that modifies the tree to invalidate iterators.

Reply:
Resolved in https://github.com/ETCLabs/sACN/pull/32/commits/9a88cc62e4c3fcb6011291027e0f8274c5237b9a.